### DISCUSSION TITLE
mlflow: backend-store-uri를 gcs나 big query로 한다면?

### DISCUSSION THREAD

**Discussion by dnlpys** [minhyeok0328/next-loop#70] [Ideas]
### 1. `-backend-store-uri`를 GCS로 설정하는 경우

GCS는 MLflow의 **Artifact Storage**로 사용하는 것이 일반적이며, 이를 `--backend-store-uri`로 설정할 수는 없지만, 메타데이터나 실험 정보를 직접 GCS에 저장할 수 있습니다. 다만 GCS는 파일 기반 저장소이므로, 실험 메타데이터를 직접 저장하기에는 효율적이지 않습니다. 대안으로, **Cloud SQL (MySQL이나 PostgreSQL)**을 백엔드로 사용하는 것을 추천드립니다. Cloud SQL과 GCS의 구성을 함께 사용해 보세요.

### 2. `-backend-store-uri`를 BigQuery로 설정하는 경우

MLflow는 직접적으로 BigQuery를 `backend-store-uri`로 지원하지 않습니다. 그러나 MLflow 메타데이터를 외부 데이터베이스에 저장한 후, BigQuery와 연결해 조회하거나 분석하는 방법으로 활용할 수 있습니다.

### 실험 메타데이터를 Cloud SQL과 연결하는 방법 (BigQuery 연동 가능)

1. **Cloud SQL 설정**:
    - GCP Console에서 **Cloud SQL 인스턴스**를 생성하고 MySQL 혹은 PostgreSQL을 선택합니다.
    - 데이터베이스 및 사용자 계정을 설정하고 인스턴스의 IP 주소를 VM에서 접근할 수 있도록 설정합니다.
2. **Cloud SQL 데이터베이스 URI 구성**:
    - `mysql+pymysql://username:password@host/database_name` 형식으로 Cloud SQL URI를 생성합니다.
    예: `mysql+pymysql://mlflow_user:mlflow_password@34.123.45.67/mlflow_db`
    - 여기서 `username`, `password`, `host`, `database_name`은 Cloud SQL 설정에 맞게 지정하세요.
3. **MLflow Tracking 서버 실행**:
    
    ```bash
    mlflow server \
        --backend-store-uri mysql+pymysql://mlflow_user:mlflow_password@34.123.45.67/mlflow_db \
        --default-artifact-root gs://your-gcs-bucket/mlflow-artifacts \
        --host 0.0.0.0 \
        --port 5000
    
    ```
    
4. **BigQuery 연동**:
    - Cloud SQL에 저장된 MLflow 메타데이터를 BigQuery로 이동하거나 연결할 수 있습니다.
    - **Cloud SQL Federation**을 통해 BigQuery에서 Cloud SQL 테이블을 직접 쿼리하거나, 데이터 통합 파이프라인을 구축해 주기적으로 데이터를 BigQuery로 옮길 수 있습니다.

### 요약

- GCS를 `backend-store-uri`로 사용하는 것은 비효율적이므로 권장되지 않습니다.
- Cloud SQL (MySQL/PostgreSQL) 인스턴스를 `backend-store-uri`로 사용하고, BigQuery와 연동하여 메타데이터를 분석하는 방식을 추천드립니다.

**Comment by SOONSOO7** [comment]
음.. airflow만 보더라도 postgre나 mysql 안쓰니까
Do not use SQLite as metadata DB in production – it should only be used for dev/testing. We recommend using Postgres or MySQL. 
Do not use the SequentialExecutor in production. (이거는 LocalExecutor 쓰라는 것)
알람이 나오긴 하더라구요

**Comment by imella0707** [comment]
따라서 해볼 때 도움이 많이 될 것 같아요. 좋은 정보 감사합니다 ~!
